### ISSUE TITLE
PropertyChangeProvider is not triggered after calling setState

### ISSUE THREAD

**Issue by giordy16** [VGVentures/property_change_notifier#23]
This problem is very simple to reproduce. If I click on +1 button, the value of the counter is updated. Then, if I click on the setState button, the counter is reset, but if I try to increase the counter, the `PropertyChangeProvider` doesn't get triggered. I would except it does. 

To solve this i can simply move the declaration of the counter outside the `build` method, but isn't really what I am looking for, since this code is a simplification of my situation.

On my code, I have a list of items, which are initialized on the `initState` method of the screen. Then on the screen there is a FAB button which brings me to a screen to create a new item. Once I am back, I fetch again the items and call `setState` to update the screen. At this point I have the problem of the discussion. 

```
import 'package:flutter/material.dart';
import 'package:pdfx/pdfx.dart';
import 'package:property_change_notifier/property_change_notifier.dart';

void main() {
  runApp(const MaterialApp(
    home: MyApp(),
  ));
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    Counter c = Counter();

    return Scaffold(
        body: SafeArea(
            child: Center(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          CounterValueWidget(c),
          ElevatedButton(
              onPressed: () {
                c.value = (c.value ?? 0) + 1;
              },
              child: Text("+1")),
          ElevatedButton(
              onPressed: () {
                setState(() {});
              },
              child: Text("setState"))
        ],
      ),
    )));
  }
}

class Counter with PropertyChangeNotifier<String> {
  int? _value = 0;

  Counter({value}) {
    _value = value;
  }

  int? get value => _value;

  set value(int? value) {
    if (value != null) {
      _value = value;
      notifyListeners("value");
    }
  }
}

class CounterValueWidget extends StatefulWidget {
  Counter counter;

  CounterValueWidget(this.counter);

  @override
  State<StatefulWidget> createState() => _CounterValueWidgetState();
}

class _CounterValueWidgetState extends State<CounterValueWidget> {
  @override
  Widget build(BuildContext context) {
    return PropertyChangeProvider<Counter, String>(
        key: Key("value"),
        value: widget.counter,
        child: Builder(builder: (BuildContext context) {
          final c = PropertyChangeProvider.of<Counter, String>(context,
                  properties: ["value"])!
              .value;

          return Text((c.value ?? 0).toString());
        }));
  }
}
```

**Comment by martinrybak** [comment]
I strongly suggest against creating long-lived model instances inside the `build` method. The [Flutter docs](https://docs.flutter.dev/get-started/fundamentals/widgets) state "This method can potentially be called in every frame and should not have any side effects beyond building a widget." However your example did point out an issue in the code. Please try out branch `#23` to see if it behaves as you intend. Once you give the ok I will merge and release a fix. Thanks!

**Comment by giordy16** [comment]
Hi @martinrybak , first of all thanks for the quick support!

Yes you are right about the use of models inside the `build` method. That's why I decided to change my code a little bit.
The fix you have done in the branch `#23` actually solves the problem of the code I have posted. 

However, I still have a problem. On my code I am using [objectbox](https://pub.dev/packages/objectbox) as local database to save and read models. [Here](https://github.com/giordy16/PropertyChangeProviderIssue) you can find my code. These are the steps to reproduce the problem:

- create a new counter with + FAB
- click +1 on the first counter, and observe the widget updating
- create a new counter with + FAB
- click +1 on the first counter, and observe the widget does NOT update
- click +1 on the last created counter, and observe the widget updating

I have been debugging this all the morning but I can not really give me an explanation of the problem since the code is more or less simple, and if I am not using objectbox but just a local list where i just add and remove Counters is working fine.

EDIT: if I remove the `EquatableMinMax` from the model (or I don't override `operator==` and `hashCode`) it's actually working, but I don't understand why, and removing this is not the solution I am looking for

**Comment by giordy16** [comment]
Ok apparently it was a widget problem. I have set now an `UniqueKey` to each widget of the list that has `PropertyChangeProvider` and it works! 
```
          child: ListView.separated(
            itemBuilder: (context, index) {
              Counter c = counterList[index];
              return Row(
                children: [
                  Expanded(child: CounterValueWidget(c, key: UniqueKey())),   <------------
                  IconButton(
                      onPressed: () {
                        delete(c);
                        setState(() {
                          if (useObjectBox) fetchCounters();
                        });
                      },
                      icon: const Icon(Icons.delete_outline)),
                  IconButton(
                      onPressed: () {
                        increase(c);
                      },
                      icon: const Icon(Icons.plus_one))
                ],
              );
            },
            separatorBuilder: (context, index) {
              return const Divider();
            },
            itemCount: counterList.length,
          ),
```

But I still don't understand why using objectbox was giving the problem, instead using a local list was working 🤔

**Comment by martinrybak** [comment]
Glad to hear! Thanks for providing the code to make it easy to debug your issue.

**Comment by martinrybak** [comment]
Released in version 0.5.0
